packages/angular/ssr/src/routes/ng-routes.ts

```diff
@@ -0,0 +1,228 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import { APP_BASE_HREF, PlatformLocation } from '@angular/common';
+import {
+  ApplicationRef,
+  Compiler,
+  Injector,
+  createPlatformFactory,
+  platformCore,
+  ɵwhenStable as whenStable,
+  ɵConsole,
+} from '@angular/core';
+import {
+  INITIAL_CONFIG,
+  ɵINTERNAL_SERVER_PLATFORM_PROVIDERS as INTERNAL_SERVER_PLATFORM_PROVIDERS,
+} from '@angular/platform-server';
+import {
+  RedirectFunction,
+  Route,
+  Router,
+  ɵloadChildren as loadChildrenHelper,
+} from '@angular/router';
+import { Console } from '../console';
+import { AngularBootstrap, isNgModule } from '../utils/ng';
+import { joinUrlParts } from '../utils/url';
+
+/**
+ * Represents the result of processing a route.
+ */
+export interface RouteResult {
+  /**
+   * The resolved path of the route.
+   *
+   * This string represents the complete URL path for the route after it has been
+   * resolved, including any parent routes or path segments that have been joined.
+   */
+  route: string;
+
+  /**
+   * The target path for route redirection, if applicable.
+   *
+   * If this route has a `redirectTo` property in the configuration, this field will
+   * contain the full resolved URL path that the route should redirect to.
+   */
+  redirectTo?: string;
+}
+
+/**
+ * Recursively traverses the Angular router configuration to retrieve routes.
+ *
+ * Iterates through the router configuration, yielding each route along with its potential
+ * redirection or error status. Handles nested routes and lazy-loaded child routes.
+ *
+ * @param options - An object containing the parameters for traversing routes.
+ * @returns An async iterator yielding `RouteResult` objects.
+ */
+async function* traverseRoutesConfig(options: {
+  /** The array of route configurations to process. */
+  routes: Route[];
+  /** The Angular compiler used to compile route modules. */
+  compiler: Compiler;
+  /** The parent injector for lazy-loaded modules. */
+  parentInjector: Injector;
+  /** The parent route path to prefix child routes. */
+  parentRoute: string;
+  /** The base URL of the application used to resolve relative paths. */
+  baseHref: string;
```

Comment:
Is it necessary to carry the baseHref around and attach it to all the route results? Can't the base href logic just be applied to all results where you use them and the `RouteResults` are all just "relative to the app base"?

Reply:
Great idea. I think it should be doubled. I'll give it a try tomorrow.